spec/haskell/src/SEL4/Object/SchedContext.lhs

```diff
@@ -149,22 +152,28 @@ This module uses the C preprocessor to select a target architecture.
 >     sc <- getSchedContext scPtr
 >     return $ scRefillCount sc == 0
 
-> mapScPtr :: PPtr SchedContext -> (SchedContext -> a) -> Kernel a
-> mapScPtr scPtr f = do
->     sc <- getSchedContext scPtr
+> readMapScPtr :: PPtr SchedContext -> (SchedContext -> a) -> KernelR a
+> readMapScPtr scPtr f = do
+>     sc <- readSchedContext scPtr
 >     return $ f sc
 
+> getMapScPtr :: PPtr SchedContext -> (SchedContext -> a) -> Kernel a
+> getMapScPtr scPtr f = getsJust (readMapScPtr scPtr f)
+
 > updateScPtr :: PPtr SchedContext -> (SchedContext -> SchedContext) -> Kernel ()
 > updateScPtr scPtr f = do
 >     sc <- getSchedContext scPtr
 >     setSchedContext scPtr (f sc)
 
-> refillNext :: PPtr SchedContext -> Int -> Kernel Int
-> refillNext scPtr index = do
->     sc <- getSchedContext scPtr
+> readRefillNext :: PPtr SchedContext -> Int -> KernelR Int
+> readRefillNext scPtr index = do
+>     sc <- readSchedContext scPtr
 >     return $ if (index == scRefillMax sc - 1) then 0 else index + 1
 
-> -- Odd argument order plays well with `mapScPtr`.
+> getRefillNext :: PPtr SchedContext -> Int -> Kernel Int
+> getRefillNext scPtr index = getsJust (readRefillNext scPtr index)
```

Comment:
I think I was mostly just copying what we had in the Haskell when I did this work, though maybe not. @mktnk3 could you let me know what you'd like to do with this and with `mapScPtr`?

Reply:
For `mapScPtr`, maybe we do a JIRA issue for future cleanup and leave it to that for now?

As for having both readers and getters, it is a bit hard to say but, at this point, I think we want to be using the getter version whenever possible. So far, the reader monad is used in a minimum way, and that seems to be more efficient in general, probably because we have a lot more infrastructure for `nondet_monad`. Technically, I suppose we could replace getters with readers altogether, and that would be more consistent, but I am not sure if we gain much from it immediately given how things are set up right now. Maybe one day.